vpn/client.go

```diff
@@ -32,6 +32,11 @@ type tunnel struct {
 type vpnClient interface {
 	Start() error
 	Dial() (net.Conn, error)
+	WithContext(context.Context) vpnClient
```

Comment:
If possible, I'd rather pass the context explicitly as argument rather than doing what you are doing here, which strikes me as storing the context inside a `vpnClient`. Please, see https://pkg.go.dev/context for more.